test/distributed/fsdp/test_fsdp_grad_acc.py

```diff
@@ -0,0 +1,146 @@
+    sys.exit(0)
+
+if TEST_WITH_DEV_DBG_ASAN:
+    print(
+        "Skip dev-asan as torch + multiprocessing spawn have known issues",
+        file=sys.stderr,
+    )
+    sys.exit(0)
+
+
+class TestGradAcc(FSDPTest):
+    """Tests ``FullyShardedDataParallel``'s gradient accumulation via its
+    ``no_sync()`` context manager."""
+
+    def _test_no_sync(self, batch_dim, num_iters_to_acc):
+        """
+        Tests ``no_sync()`` by comparing a run that trains sequentially through
+        some batches while accumulating gradients with a run that trains on the
+        concatenation of those batches in a single iteration. The number of
+        batches, i.e. the number of iterations for which to accumulate
+        gradients, is given by ``num_iters_to_acc``.
+        """
+        # Use double precision to avoid floating point drift
+        torch.set_default_dtype(torch.float64)
+
+        # Initialize the FSDP model
+        group = dist.distributed_c10d._get_default_group()
+        fsdp_model: FSDP = self._get_wrapped_model(
+            group, cuda_first=False, add_bn=False,
+        )  # disable BN since the test uses variable batch size
+        fsdp_model.eval()  # disable dropout
+        device = torch.device("cuda")
+
+        # Generate the sequence of batches, each containing the same data but
+        # permuted
+        def permute_tensor(x: torch.Tensor):
+            return x.view(-1)[torch.randperm(x.numel())].view_as(x)
+
+        batch: Tuple[torch.Tensor, ...] = fsdp_model.module.get_input(device)
+        batches: List[Tuple[torch.Tensor, ...]] = [batch]
+        for _ in range(num_iters_to_acc - 1):
+            batches.append(tuple(permute_tensor(t) for t in batch))
+        for (batch1, batch2) in itertools.combinations(batches, r=2):
+            for t1, t2 in zip(batch1, batch2):
+                assert not torch.all(t1 == t2)
+
+        # Concatenate the batches along the given batch dimension
+        concat_batch: Tuple[torch.Tensor, ...] = tuple(
+            torch.cat(ts, dim=batch_dim) for ts in zip(*batches)
+        )
+
+        # Establish reference gradients using the concatenated batch
+        fsdp_model.zero_grad()
+        output = fsdp_model(*concat_batch)
+        ref_loss = fsdp_model.module.get_loss(concat_batch, output)
+        ref_loss.backward()
+        ref_grads = [p.grad.detach().clone() for p in fsdp_model.parameters()]
+
+        # Compute the gradients by accumulating via `no_sync()`
+        fsdp_model.zero_grad()
+        losses = []
+        with fsdp_model.no_sync():
+            for batch in batches[:-1]:  # accumulate for all but the last batch
+                output = fsdp_model(*batch)
+                loss = fsdp_model.module.get_loss(batch, output)
+                loss.backward()
+                losses.append(loss)
+        output = fsdp_model(*batches[-1])
+        loss = fsdp_model.module.get_loss(batches[-1], output)
```

Comment:
Should we run an optimizer.step() after no_sync() is exited just to make sure things work fine?

Reply:
Added.